src/data/setoid.lean

```diff
@@ -0,0 +1,580 @@
+definitional equalities (a similar example is `filter.lattice.complete_lattice` in 
+`order/filter/basic.lean`). This does not save space, however, and is less clear.
+
+Partitions are not defined as a separate structure here; users are encouraged to
+reason about them using the existing `setoid` and its infrastructure.
+
+Eventually there should be a proof here that Bell numbers are the number of possible
+partitions of a finite set.
+
+## References
+
+* https://en.wikipedia.org/wiki/Bell_triangle
+
+## Tags
+
+setoid, equivalence, iseqv, relation, equivalence relation, partition, equivalence 
+class, Bell number, Bell triangle
+-/
+variables {α : Type*} {β : Type*}
+
+open lattice 
+
+namespace setoid
+
+/-- A version of setoid.r that takes the equivalence relation as an explicit argument. -/
+def rel (r : setoid α) := @setoid.r _ r
+
+@[extensionality] lemma ext' {r s : setoid α} (H : ∀ a b, r.rel a b ↔ s.rel a b) : 
+  r = s := ext H
+
+lemma ext_iff {r s : setoid α} : r = s ↔ ∀ a b, r.rel a b ↔ s.rel a b :=
+⟨λ h a b, h ▸ iff.rfl, ext'⟩
+
+/-- Defining '≤' for equivalence relations. -/
+instance : has_le (setoid α) := ⟨λ r s, ∀ x y, r.rel x y → s.rel x y⟩
+
+theorem le_def {r s : setoid α} : r ≤ s ↔ ∀ {x y}, r.rel x y → s.rel x y := iff.rfl
+
+@[refl] lemma refl' (r : setoid α) (x) : r.rel x x := r.2.1 x
+@[symm] lemma symm' (r : setoid α) : ∀ {x y}, r.rel x y → r.rel y x := λ _ _ h, r.2.2.1 h
+@[trans] lemma trans' (r : setoid α) : ∀ {x y z}, r.rel x y → r.rel y z → r.rel x z := 
+λ  _ _ _ hx hy, r.2.2.2 hx hy
+
+/-- The kernel of a function is an equivalence relation. -/
+def ker (f : α → β) : setoid α :=
+⟨λ x y, f x = f y, ⟨λ x, rfl, λ _ _ h, h.symm, λ _ _ _ h1 h2, h1.trans h2⟩⟩ 
+
+/-- The kernel of the quotient map induced by an equivalence relation r equals r. -/
+@[simp] lemma ker_mk_eq (r : setoid α) : ker (@quotient.mk _ r) = r :=
+ext' $ λ x y, quotient.eq
+
+/-- The infimum of two equivalence relations. -/
+instance : has_inf (setoid α) :=
+⟨λ r s, ⟨λ x y, r.rel x y ∧ s.rel x y, ⟨λ x, ⟨r.refl' x, s.refl' x⟩, 
+ λ _ _ h, ⟨r.symm' h.1, s.symm' h.2⟩, 
+ λ _ _ _ h1 h2, ⟨r.trans' h1.1 h2.1, s.trans' h1.2 h2.2⟩⟩⟩⟩
+
+/-- The infimum of 2 equivalence relations r and s is the same relation as the infimum 
+    of the underlying binary operations. -/
+lemma inf_def {r s : setoid α} : (r ⊓ s).rel = r.rel ⊓ s.rel := rfl 
+
+theorem inf_iff_and {r s : setoid α} {x y} : 
+  (r ⊓ s).rel x y ↔ r.rel x y ∧ s.rel x y := iff.rfl
+
+/-- The infimum of a set of equivalence relations. -/
+instance : has_Inf (setoid α) :=
+⟨λ S, ⟨λ x y, ∀ r ∈ S, rel r x y, 
+⟨λ x r hr, r.refl' x, λ _ _ h r hr, r.symm' $ h r hr, 
+ λ _ _ _ h1 h2 r hr, r.trans' (h1 r hr) $ h2 r hr⟩⟩⟩
+
+/-- The underlying binary operation of the infimum of a set of equivalence relations
+    is the infimum of the set's image under the map to the underlying binary operation. -/
+theorem Inf_def {s : set (setoid α)} : (Inf s).rel = Inf (rel '' s) :=
+begin
```

Comment:
This can be proved with `by { ext, simp only [Inf_image, infi_apply, infi_Prop_eq], refl }